MapboxCoreNavigation/TunnelIntersectionManager.swift

```diff
@@ -0,0 +1,146 @@
+        
+        guard let currentIntersection = routeProgress.currentLegProgress.currentStepProgress.currentIntersection else {
+            return false
+        }
+        
+        if let classes = currentIntersection.outletRoadClasses, classes.contains(.tunnel) {
+            return true
+        }
+        
+        return userWithinTunnelEntranceRadius(at: location, routeProgress: routeProgress)
+    }
+    
+    /**
+     Given a user's current location and the route progress,
+     detects whether the upcoming intersection contains a tunnel road class, and
+     returns a Boolean whether they are within the minimum radius of a tunnel entrance.
+     */
+    @objc public func userWithinTunnelEntranceRadius(at location: CLLocation, routeProgress: RouteProgress) -> Bool {
+        // Ensure the upcoming intersection is a tunnel intersection
+        // OR the location speed is either at least 5 m/s or is considered a bad location update
+        guard let upcomingIntersection = routeProgress.currentLegProgress.currentStepProgress.upcomingIntersection,
+            let roadClasses = upcomingIntersection.outletRoadClasses, roadClasses.contains(.tunnel),
+            (location.speed >= RouteControllerMinimumSpeedAtTunnelEntranceRadius || !location.isQualified) else {
+                return false
+        }
+        
+        // Distance to the upcoming tunnel entrance
+        guard let distanceToTunnelEntrance = routeProgress.currentLegProgress.currentStepProgress.userDistanceToUpcomingIntersection else { return false }
+        
+        return distanceToTunnelEntrance < RouteControllerMinimumDistanceToTunnelEntrance
+    }
+    
+    @objc public func enableTunnelAnimation(for manager: CLLocationManager,
+                                        routeController: RouteController,
+                                          routeProgress: RouteProgress,
+                                               callback: RouteControllerSimulationCompletionBlock?) {
+        guard !isAnimationEnabled else { return }
+        
+        isAnimationEnabled = true
+        
+        self.animatedLocationManager = SimulatedLocationManager(routeProgress: routeProgress)
+        self.animatedLocationManager?.delegate = routeController
+        self.animatedLocationManager?.routeProgress = routeProgress
+        self.animatedLocationManager?.startUpdatingHeading()
+        
+        callback?(isAnimationEnabled, self.animatedLocationManager!)
+    }
+    
+    @objc public func suspendTunnelAnimation(for manager: CLLocationManager,
+                                             at location: CLLocation,
+                                         routeController: RouteController,
+                                                callback: RouteControllerSimulationCompletionBlock?) {
+        
+        guard isAnimationEnabled else { return }
+        
+        // Disable the tunnel animation after at least 3 bad location updates.
```

Comment:
It think it might make more sense to disable tunnel animation after 3 _good_ location updates. If we get 3 bad location updates in a row, we should not do anything IMO.

Reply:
@bsudekum just for the record, this might delay style switches and delay in puck movements in case the driver exits the tunnel very quickly. (PS: I will still proceed with the suggested changes.)